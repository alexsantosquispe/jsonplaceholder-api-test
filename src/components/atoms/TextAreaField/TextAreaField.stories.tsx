import { Meta, StoryObj } from '@storybook/react';
import { Control, FieldValues, useForm } from 'react-hook-form';

import { TextAreaField } from './TextAreaField';
import { Wrapper } from '../../../utils/stories-utils';

const meta = {
  title: 'Atoms/TextAreaField',
  component: TextAreaField,
  parameters: {
    layout: 'centered'
  }
} satisfies Meta<typeof TextAreaField>;

export default meta;

type Story = StoryObj<typeof TextAreaField>;

export const Default: Story = () => {
  const { control } = useForm();

  return (
    <div className="grid grid-cols-4 gap-4">
      <Wrapper>
        <TextAreaField
          label="Textarea label"
          name="textarea-1"
          control={control as Control<FieldValues>}
          placeholder="Enter text..."
        />
      </Wrapper>
      <Wrapper>
        <TextAreaField
          label="Textarea required"
          name="textarea-2"
          control={control as Control<FieldValues>}
          placeholder="Enter text..."
          isRequired
        />
      </Wrapper>
      <Wrapper>
        <TextAreaField
          control={control as Control<FieldValues>}
          label="Textarea disabled"
          name="textarea-3"
          placeholder="Enter text..."
          isDisabled
        />
      </Wrapper>
      <Wrapper>
        <TextAreaField
          control={control as Control<FieldValues>}
          label="Textarea with error message"
          name="textarea-4"
          placeholder="Enter text..."
          errorMessage="This field has an error"
        />
      </Wrapper>
    </div>
  );
};

Default.args = {};
