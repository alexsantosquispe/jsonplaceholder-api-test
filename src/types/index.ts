import { Control, FieldValues } from 'react-hook-form';

export interface SvgIconProps {
  className?: string;
}

export interface InputProps {
  label: string;
  name: string;
  control: Control<FieldValues>;
  placeholder?: string;
  type?: string;
  errorMessage?: string;
  isDisabled?: boolean;
  isRequired?: boolean;
}

export type TextAreaProps = Omit<InputProps, 'type'> & {
  rows?: number;
  cols?: number;
};
