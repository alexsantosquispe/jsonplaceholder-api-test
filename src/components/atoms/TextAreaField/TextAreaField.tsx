import cn from 'clsx';
import { Controller } from 'react-hook-form';
import { twMerge } from 'tailwind-merge';

import {
  INPUT_FIELD_ERROR,
  INPUT_FIELD_STYLES
} from '../../../constants/styles';
import { TextAreaProps } from '../../../types';

export const TextAreaField = ({
  label,
  name,
  control,
  placeholder,
  rows = 6,
  cols = 50,
  isDisabled,
  errorMessage = '',
  isRequired = false
}: TextAreaProps) => {
  return (
    <Controller
      name={name}
      control={control}
      render={({ field: { value, onChange } }) => {
        return (
          <div className="flex flex-col gap-y-0.5">
            <label htmlFor={name} className="mb-0.5 text-sm">
              {label}
              {isRequired && <span className="text-rose-600">&nbsp;*</span>}
            </label>
            <textarea
              id={name}
              name={name}
              placeholder={placeholder}
              disabled={isDisabled}
              onChange={onChange}
              value={value}
              rows={rows}
              cols={cols}
              className={twMerge(
                INPUT_FIELD_STYLES,
                'resize-none',
                cn({
                  'border-rose-600 dark:border-rose-400': !!errorMessage
                })
              )}
            />
            {errorMessage && (
              <span className={INPUT_FIELD_ERROR}>{errorMessage}</span>
            )}
          </div>
        );
      }}
    />
  );
};
