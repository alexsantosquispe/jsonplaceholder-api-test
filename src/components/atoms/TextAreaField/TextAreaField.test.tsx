import { render, screen } from '@testing-library/react';
import { Control, FieldValues, useForm } from 'react-hook-form';

import { TextAreaField } from './TextAreaField';
import { TextAreaProps } from '../../../types';

const TextAreaWrapper = (props: Omit<TextAreaProps, 'control'>) => {
  const { control } = useForm();
  return <TextAreaField {...props} control={control as Control<FieldValues>} />;
};

describe('TextAreaField', () => {
  const errorMessage = 'Field with error';
  const props = {
    label: 'Input text',
    name: 'firstName',
    placeholder: 'Enter text...'
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render the component correctly', () => {
    let component = render(<TextAreaWrapper {...props} />);
    expect(component).toMatchSnapshot();

    component = render(<TextAreaWrapper {...props} isRequired />);
    expect(component).toMatchSnapshot();
  });

  it('should show required indicator when isRequired is true', () => {
    render(<TextAreaWrapper {...props} isRequired />);

    expect(screen.getByText('*')).toBeInTheDocument();
  });

  it('should the input be disabled', () => {
    render(<TextAreaWrapper {...props} isDisabled />);

    const textAreaDisabled = screen.getByPlaceholderText(props.placeholder);

    expect(textAreaDisabled).toBeInTheDocument();

    expect(textAreaDisabled).toBeDisabled();
  });

  it('should display error message and error styles', () => {
    render(<TextAreaWrapper {...props} errorMessage={errorMessage} />);

    const textArea = screen.getByPlaceholderText(props.placeholder);

    expect(textArea).toBeInTheDocument();

    expect(textArea).toHaveClass('border-rose-600');

    expect(screen.getByText(errorMessage)).toBeInTheDocument();
  });
});
