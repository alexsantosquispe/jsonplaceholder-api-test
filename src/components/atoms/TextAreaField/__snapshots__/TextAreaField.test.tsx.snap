// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextAreaField should render the component correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col gap-y-0.5"
      >
        <label
          class="mb-0.5 text-sm"
          for="firstName"
        >
          Input text
        </label>
        <textarea
          class="w-full rounded-lg border border-gray-200 px-4 py-2 text-sm dark:border-white/30 focus:ring-primary/60 focus:border-primary/60 focus:ring-1 focus:outline-none dark:focus:border-white/65 dark:focus:ring-white/65 resize-none"
          cols="50"
          id="firstName"
          name="firstName"
          placeholder="Enter text..."
          rows="6"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col gap-y-0.5"
    >
      <label
        class="mb-0.5 text-sm"
        for="firstName"
      >
        Input text
      </label>
      <textarea
        class="w-full rounded-lg border border-gray-200 px-4 py-2 text-sm dark:border-white/30 focus:ring-primary/60 focus:border-primary/60 focus:ring-1 focus:outline-none dark:focus:border-white/65 dark:focus:ring-white/65 resize-none"
        cols="50"
        id="firstName"
        name="firstName"
        placeholder="Enter text..."
        rows="6"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`TextAreaField should render the component correctly 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col gap-y-0.5"
      >
        <label
          class="mb-0.5 text-sm"
          for="firstName"
        >
          Input text
        </label>
        <textarea
          class="w-full rounded-lg border border-gray-200 px-4 py-2 text-sm dark:border-white/30 focus:ring-primary/60 focus:border-primary/60 focus:ring-1 focus:outline-none dark:focus:border-white/65 dark:focus:ring-white/65 resize-none"
          cols="50"
          id="firstName"
          name="firstName"
          placeholder="Enter text..."
          rows="6"
        />
      </div>
    </div>
    <div>
      <div
        class="flex flex-col gap-y-0.5"
      >
        <label
          class="mb-0.5 text-sm"
          for="firstName"
        >
          Input text
          <span
            class="text-rose-600"
          >
             *
          </span>
        </label>
        <textarea
          class="w-full rounded-lg border border-gray-200 px-4 py-2 text-sm dark:border-white/30 focus:ring-primary/60 focus:border-primary/60 focus:ring-1 focus:outline-none dark:focus:border-white/65 dark:focus:ring-white/65 resize-none"
          cols="50"
          id="firstName"
          name="firstName"
          placeholder="Enter text..."
          rows="6"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col gap-y-0.5"
    >
      <label
        class="mb-0.5 text-sm"
        for="firstName"
      >
        Input text
        <span
          class="text-rose-600"
        >
           *
        </span>
      </label>
      <textarea
        class="w-full rounded-lg border border-gray-200 px-4 py-2 text-sm dark:border-white/30 focus:ring-primary/60 focus:border-primary/60 focus:ring-1 focus:outline-none dark:focus:border-white/65 dark:focus:ring-white/65 resize-none"
        cols="50"
        id="firstName"
        name="firstName"
        placeholder="Enter text..."
        rows="6"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
